#Loading the data:
Sometimes data fails to load with default encoding that is "UTF-8", If file was created in older system it might have "Latin-1". We also have to look for separators that are used in our files, like its comma(','), semi-colon(';') or sometimes its simply tab('\t')


In [2]:
import pandas as pd
try:
  df = pd.read_csv('/content/Messy_Employee_dataset.csv')
except UnicodeDecodeError:
  df = pd.read_csv('/content/Messy_Employee_dataset.csv', sep=',', encoding='latin-1')

We can also explicity tell reader method about datatypes of columns to avoid silent typecasting

In [3]:
df = pd.read_csv('/content/Messy_Employee_dataset.csv', dtype={'price':float, 'quantity':'Int64'})

If you are working with big data its always good to load just first few rows to take a look at data, because loading whole data might take a lot of time.

In [4]:
df = pd.read_csv('/content/Messy_Employee_dataset.csv', nrows=10)

In [5]:
df = pd.read_csv('/content/Messy_Employee_dataset.csv')

In [10]:
df

Employee_ID First_Name Last_Name   Age   Department_Region    Status  \
0        EMP1000        Bob     Davis  25.0   DevOps-California    Active   
1        EMP1001        Bob     Brown   NaN       Finance-Texas    Active   
2        EMP1002      Alice     Jones   NaN        Admin-Nevada   Pending   
3        EMP1003        Eva     Davis  25.0        Admin-Nevada  Inactive   
4        EMP1004      Frank  Williams  25.0  Cloud Tech-Florida    Active   
...          ...        ...       ...   ...                 ...       ...   
1015     EMP2015      David    Miller  30.0       HR-California    Active   
1016     EMP2016      David   Johnson  30.0    Cloud Tech-Texas  Inactive   
1017     EMP2017    Charlie  Williams  40.0    Finance-New York    Active   
1018     EMP2018      Alice    Garcia  30.0          HR-Florida  Inactive   
1019     EMP2019      Heidi     Jones  30.0     DevOps-Illinois   Pending   

       Join_Date     Salary                         Email       Phone  \
0       4/2/2021   59767.65         bob.davis@example.com -1651623197   
1      7/10/2020   65304.66         bob.brown@example.com -1898471390   
2      12/7/2023   88145.90       alice.jones@example.com -5596363211   
3     11/27/2021   69450.99         eva.davis@example.com -3476490784   
4       1/5/2022  109324.61    frank.williams@example.com -1586734256   
...          ...        ...                           ...         ...   
1015   8/19/2023        NaN      david.miller@example.com -3546212759   
1016   11/7/2021  100215.06     david.johnson@example.com -2508261122   
1017   10/4/2023  114587.11  charlie.williams@example.com -1261632487   
1018  12/16/2024   71318.79      alice.garcia@example.com -8995729892   
1019   2/22/2021   77764.24       heidi.jones@example.com -7629745492   

     Performance_Score  Remote_Work  
0              Average         True  
1            Excellent         True  
2                 Good         True  
3                 Good         True  
4                 Poor        False  
...                ...          ...  
1015              Good         True  
1016              Good         True  
1017           Average        False  
1018              Good         True  
1019           Average        False  

[1020 rows x 12 columns]

# Inspection Of Data

In [6]:
df.head()

Employee_ID First_Name Last_Name   Age   Department_Region    Status  \
0     EMP1000        Bob     Davis  25.0   DevOps-California    Active   
1     EMP1001        Bob     Brown   NaN       Finance-Texas    Active   
2     EMP1002      Alice     Jones   NaN        Admin-Nevada   Pending   
3     EMP1003        Eva     Davis  25.0        Admin-Nevada  Inactive   
4     EMP1004      Frank  Williams  25.0  Cloud Tech-Florida    Active   

    Join_Date     Salary                       Email       Phone  \
0    4/2/2021   59767.65       bob.davis@example.com -1651623197   
1   7/10/2020   65304.66       bob.brown@example.com -1898471390   
2   12/7/2023   88145.90     alice.jones@example.com -5596363211   
3  11/27/2021   69450.99       eva.davis@example.com -3476490784   
4    1/5/2022  109324.61  frank.williams@example.com -1586734256   

  Performance_Score  Remote_Work  
0           Average         True  
1         Excellent         True  
2              Good         True  
3              Good         True  
4              Poor        False

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1020 entries, 0 to 1019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Employee_ID        1020 non-null   object 
 1   First_Name         1020 non-null   object 
 2   Last_Name          1020 non-null   object 
 3   Age                809 non-null    float64
 4   Department_Region  1020 non-null   object 
 5   Status             1020 non-null   object 
 6   Join_Date          1020 non-null   object 
 7   Salary             996 non-null    float64
 8   Email              1020 non-null   object 
 9   Phone              1020 non-null   int64  
 10  Performance_Score  1020 non-null   object 
 11  Remote_Work        1020 non-null   bool   
dtypes: bool(1), float64(2), int64(1), object(8)
memory usage: 88.8+ KB


We can see clealy that age column shows 809 values, but rest are empty that is significant data missing, same goes for Salary column that too has many missing values

In [9]:
df.duplicated().sum()

np.int64(0)

This shows 0 that means our data set contains zero duplicates

In [10]:
df['Employee_ID'].unique()

array(['EMP1000', 'EMP1001', 'EMP1002', ..., 'EMP2017', 'EMP2018',
       'EMP2019'], dtype=object)

In [11]:
df.describe()

Age         Salary         Phone
count  809.000000     996.000000  1.020000e+03
mean    32.484549   85155.056396 -4.942253e+09
std      5.656860   19873.727918  2.817326e+09
min     25.000000   50047.320000 -9.994973e+09
25%     25.000000   68392.487500 -7.341992e+09
50%     30.000000   85547.870000 -4.943997e+09
75%     40.000000  100974.027500 -2.520391e+09
max     40.000000  119971.650000 -3.896086e+06

#Standardizing, and handling missing values

In [21]:
df.columns = df.columns.str.lower().str.replace(' ','_').str.strip()
# Before: ['Employee_ID', 'First_Name', 'Phone']
# After: ['employee_id', 'first_name', 'phone']

In [23]:
#setting employee_id as index
df.set_index('employee_id', inplace=True)

In [24]:
df

first_name last_name   age   department_region    status  \
employee_id                                                            
EMP1000            Bob     Davis  25.0   DevOps-California    Active   
EMP1001            Bob     Brown   NaN       Finance-Texas    Active   
EMP1002          Alice     Jones   NaN        Admin-Nevada   Pending   
EMP1003            Eva     Davis  25.0        Admin-Nevada  Inactive   
EMP1004          Frank  Williams  25.0  Cloud Tech-Florida    Active   
...                ...       ...   ...                 ...       ...   
EMP2015          David    Miller  30.0       HR-California    Active   
EMP2016          David   Johnson  30.0    Cloud Tech-Texas  Inactive   
EMP2017        Charlie  Williams  40.0    Finance-New York    Active   
EMP2018          Alice    Garcia  30.0          HR-Florida  Inactive   
EMP2019          Heidi     Jones  30.0     DevOps-Illinois   Pending   

              join_date     salary                         email       phone  \
employee_id                                                                    
EMP1000        4/2/2021   59767.65         bob.davis@example.com -1651623197   
EMP1001       7/10/2020   65304.66         bob.brown@example.com -1898471390   
EMP1002       12/7/2023   88145.90       alice.jones@example.com -5596363211   
EMP1003      11/27/2021   69450.99         eva.davis@example.com -3476490784   
EMP1004        1/5/2022  109324.61    frank.williams@example.com -1586734256   
...                 ...        ...                           ...         ...   
EMP2015       8/19/2023        NaN      david.miller@example.com -3546212759   
EMP2016       11/7/2021  100215.06     david.johnson@example.com -2508261122   
EMP2017       10/4/2023  114587.11  charlie.williams@example.com -1261632487   
EMP2018      12/16/2024   71318.79      alice.garcia@example.com -8995729892   
EMP2019       2/22/2021   77764.24       heidi.jones@example.com -7629745492   

            performance_score  remote_work  
employee_id                                 
EMP1000               Average         True  
EMP1001             Excellent         True  
EMP1002                  Good         True  
EMP1003                  Good         True  
EMP1004                  Poor        False  
...                       ...          ...  
EMP2015                  Good         True  
EMP2016                  Good         True  
EMP2017               Average        False  
EMP2018                  Good         True  
EMP2019               Average        False  

[1020 rows x 11 columns]

In [26]:
df['phone'] = df['phone'].astype(str)

In [27]:
df['phone']

employee_id
EMP1000    -1651623197
EMP1001    -1898471390
EMP1002    -5596363211
EMP1003    -3476490784
EMP1004    -1586734256
              ...     
EMP2015    -3546212759
EMP2016    -2508261122
EMP2017    -1261632487
EMP2018    -8995729892
EMP2019    -7629745492
Name: phone, Length: 1020, dtype: object

Handling missing values

In [29]:
#Removing records where salary is empty as there are very few records
df.dropna(subset=['salary'], inplace=True)

In [32]:
# We are going to fill missing values for age with median of age column as there a lot of records
med_val = df['age'].median()
df['age'].fillna(med_val, inplace=True)

/tmp/ipython-input-2660213948.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['age'].fillna(med_val, inplace=True)


In [33]:
df['age'].isnull().sum()

np.int64(0)

In [34]:
df['salary'].isnull().sum()

np.int64(0)

#Standardizing values making them consistent

In [37]:
#We need to fix phone numbers
df['phone'] = df['phone'].str.replace(r'\D', '', regex=True)
df['phone'] = df['phone'].str.slice(-10)
df['phone']


employee_id
EMP1000    1651623197
EMP1001    1898471390
EMP1002    5596363211
EMP1003    3476490784
EMP1004    1586734256
              ...    
EMP2014    2470739200
EMP2016    2508261122
EMP2017    1261632487
EMP2018    8995729892
EMP2019    7629745492
Name: phone, Length: 996, dtype: object

In [38]:
df['department_region']

employee_id
EMP1000     DevOps-California
EMP1001         Finance-Texas
EMP1002          Admin-Nevada
EMP1003          Admin-Nevada
EMP1004    Cloud Tech-Florida
                  ...        
EMP2014        Finance-Nevada
EMP2016      Cloud Tech-Texas
EMP2017      Finance-New York
EMP2018            HR-Florida
EMP2019       DevOps-Illinois
Name: department_region, Length: 996, dtype: object

In [40]:
#Dividing department_region into department and region
df[['department', 'region']] = df['department_region'].str.split('-', expand=True)

In [41]:
df.drop('department_region', axis=1, inplace=True)

In [42]:
df['join_date']

employee_id
EMP1000      4/2/2021
EMP1001     7/10/2020
EMP1002     12/7/2023
EMP1003    11/27/2021
EMP1004      1/5/2022
              ...    
EMP2014     5/16/2021
EMP2016     11/7/2021
EMP2017     10/4/2023
EMP2018    12/16/2024
EMP2019     2/22/2021
Name: join_date, Length: 996, dtype: object

In [43]:
#Converting data from str obj to proper pandas datetime obj
df['join_date'] = pd.to_datetime(df['join_date'], errors='coerce')

In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 996 entries, EMP1000 to EMP2019
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   first_name         996 non-null    object        
 1   last_name          996 non-null    object        
 2   age                996 non-null    float64       
 3   status             996 non-null    object        
 4   join_date          996 non-null    datetime64[ns]
 5   salary             996 non-null    float64       
 6   email              996 non-null    object        
 7   phone              996 non-null    object        
 8   performance_score  996 non-null    object        
 9   remote_work        996 non-null    bool          
 10  department         996 non-null    object        
 11  region             996 non-null    object        
dtypes: bool(1), datetime64[ns](1), float64(2), object(8)
memory usage: 126.6+ KB


#Exporting The cleaned dataset

In [45]:
df.to_csv('cleaned_employee_datase.csv', index=True)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')